segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorage.java

```diff
@@ -0,0 +1,194 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.chunklayer;
+
+import java.io.InputStream;
+
+/**
+ * Defines an abstraction for Permanent Storage.
+ * Note: not all operations defined here are needed.
+ * <div>
+ * Below are minimum requirements that any implementation must provide.
+ * Note that it is the responsibility of storage provider specific implementation to make sure following guarantees are
+ * provided even though underlying storage may not provide all primitives or guarantees.
+ * <ul>
+ * <li>Once an operation is executed and acknowledged as successful then the effects must be permanent and consistent
+ * (as opposed to eventually consistent)</li>
+ * <li>{@link ChunkStorage#create(String)}  and {@link ChunkStorage#delete(ChunkHandle)} are not idempotent.</li>
+ * <li>{@link ChunkStorage#exists(String)} and {@link ChunkStorage#getInfo(String)} must reflect effects of most recent
+ * operation performed.</li>
+ * </ul>
+ * There is no need to implement any special logic to handle concurrent access to the underlying objects/files.
+ * </div>
+ * <div>
+ * There are a few different capabilities that ChunkStorage may provide.
+ * <ul>
+ * <li> Does {@link ChunkStorage} support appending to existing chunks?
+ * This is indicated by {@link ChunkStorage#supportsAppend()}. For example S3 compatible Chunk Storage this would return false. </li>
+ * <li> Does {@link ChunkStorage}  support for concatenating chunks? This is indicated by {@link ChunkStorage#supportsConcat()}.
+ * If this is true then concat operation concat will be invoked otherwise append functionality is invoked.</li>
+ * <li>In addition {@link ChunkStorage} may provide ability to truncate chunks at given offsets (either at front end or at tail end).
+ * This is indicated by {@link ChunkStorage#supportsTruncation()}. </li>
+ * </ul>
+ * There are some obvious constraints - If ChunkStorage supports concat but not natively then it must support append .
+ *
+ * For concats, {@link ChunkStorage} supports both native and append, ChunkManager will invoke appropriate method depending
+ * on size of target and source chunks. (Eg. ECS)
+ * </div>
+ * It is recommended that the implementations should extend {@link BaseChunkStorage}.
+ */
+public interface ChunkStorage extends AutoCloseable {
```

Comment:
- `Storage` is the interface that is already defined and used by rest of the segment store. This change does not aim to change that interface. This interface has an abstraction of segments and operates at that level of segments.
- `ChunkStorage` is a new interface which operates at individual chunks (CRDL). We wanted this to be simple interface. Implementations do not need to worry about fencing or how  these chunks are put together to form a segment. Consequently the methods and their behavior is different. Chunk is therefore not same as segment.
- We need something that translates calls at segment level (`Storage`) to calls on individual chunks (`ChunkStorage`) and also manages all these chunks, fencing, rolling etc etc . This is exactly what `ChunkManager` does. It is lot lot more than just `adapter` or `wrapper`.

Reply:
The problem isn't keeping the `Storage` interface as is, but instead that the abstractions don't quite match here, as I see it:

- `ChunkManager` should be called `ChunkStorage` instead
- `ChunkStorage` should be called something else that reflects the fact that it operates on chunks individually, e.g., `ChunkHandler` since it handles operations on chunks.

Reply:
The original names were

- `ChunkStorageManager`
- `ChunkStorageProvider`

Alternative names are 

- `ChunkStorage`
- 'ChunkProvider`, `ChunkHandler` (But Handler has connotation of handling event) ,